commons/src/quote.rs

```diff
@@ -11,11 +11,103 @@ use lb_clmm::{
 use std::collections::HashMap;
 
 #[derive(Debug)]
-pub struct SwapQuote {
+pub struct SwapExactInQuote {
     pub amount_out: u64,
     pub fee: u64,
 }
 
+#[derive(Debug)]
+pub struct SwapExactOutQuote {
+    pub amount_in: u64,
+    pub fee: u64,
+}
+
+pub fn quote_exact_out(
+    lb_pair_pubkey: Pubkey,
+    lb_pair: &LbPair,
+    mut amount_out: u64,
+    swap_for_y: bool,
+    bin_arrays: HashMap<Pubkey, BinArray>,
+    bitmap_extension: Option<&BinArrayBitmapExtension>,
+    current_timestamp: u64,
+) -> Result<SwapExactOutQuote> {
+    let mut lb_pair = *lb_pair;
+    lb_pair.update_references(current_timestamp as i64)?;
+
+    let mut total_amount_in: u64 = 0;
+    let mut total_fee: u64 = 0;
+
+    while amount_out > 0 {
+        let active_bin_array_pubkey = get_bin_array_pubkeys_for_swap(
+            lb_pair_pubkey,
+            &lb_pair,
+            bitmap_extension,
+            swap_for_y,
+            1,
+        )?
+        .pop()
+        .context("Pool out of liquidity")?;
+
+        let mut active_bin_array = bin_arrays
+            .get(&active_bin_array_pubkey)
+            .cloned()
+            .context("Active bin array not found")?;
+
+        loop {
+            if active_bin_array
+                .is_bin_id_within_range(lb_pair.active_id)
+                .is_err()
+                || amount_out == 0
+            {
+                break;
+            }
+
+            lb_pair.update_volatility_accumulator()?;
+
+            let active_bin = active_bin_array.get_bin_mut(lb_pair.active_id)?;
+            let price = active_bin.get_or_store_bin_price(lb_pair.active_id, lb_pair.bin_step)?;
+
+            if !active_bin.is_empty(!swap_for_y) {
+                let bin_max_amount_out = active_bin.get_max_amount_out(swap_for_y);
+                if amount_out >= bin_max_amount_out {
+                    let max_amount_in = active_bin.get_max_amount_in(price, swap_for_y)?;
+                    let max_fee = lb_pair.compute_fee(max_amount_in)?;
+
+                    total_amount_in = total_amount_in
+                        .checked_add(max_amount_in)
+                        .context("MathOverflow")?;
+
+                    total_fee = total_fee.checked_add(max_fee).context("MathOverflow")?;
+
+                    amount_out = amount_out
+                        .checked_sub(bin_max_amount_out)
+                        .context("MathOverflow")?;
+                } else {
+                    let amount_in = Bin::get_amount_in(amount_out, price, swap_for_y)?;
+                    let fee = lb_pair.compute_fee(amount_in)?;
+
+                    total_amount_in = total_amount_in
+                        .checked_add(amount_in)
+                        .context("MathOverflow")?;
+
+                    total_fee = total_fee.checked_add(fee).context("MathOverflow")?;
+
+                    amount_out = amount_out.checked_sub(amount_out).context("MathOverflow")?;
```

Comment:
hmm the function will always return zero?

Reply:
Fixed